lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3860,21 +3860,130 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+      else if (IID == Intrinsic::usub_sat)
+        ExtOp = OpenCLLIB::USub_sat;
+      else if (IID == Intrinsic::sadd_sat)
+        ExtOp = OpenCLLIB::SAdd_sat;
+      else
+        ExtOp = OpenCLLIB::SSub_sat;
 
-    SPIRVType *Ty = transType(II->getType());
-    std::vector<SPIRVValue *> Operands = {transValue(II->getArgOperand(0), BB),
-                                          transValue(II->getArgOperand(1), BB)};
-    return BM->addExtInst(Ty, BM->getExtInstSetId(SPIRVEIS_OpenCL), ExtOp,
-                          std::move(Operands), BB);
+      SPIRVType *Ty = transType(II->getType());
+      std::vector<SPIRVValue *> Operands = {
+          transValue(II->getArgOperand(0), BB),
+          transValue(II->getArgOperand(1), BB)};
+      return BM->addExtInst(Ty, BM->getExtInstSetId(SPIRVEIS_OpenCL), ExtOp,
+                            std::move(Operands), BB);
+    }
+
+    Type *LLVMTy = II->getType();
+    SPIRVType *Ty = transType(LLVMTy);
+    Type *BoolTy = IntegerType::getInt1Ty(M->getContext());
+    if (auto *VecTy = dyn_cast<VectorType>(LLVMTy))
+      BoolTy = VectorType::get(BoolTy, VecTy->getElementCount());
+    SPIRVType *SPVBoolTy = transType(BoolTy);
+    SPIRVValue *FirstArgVal = transValue(II->getArgOperand(0), BB);
+    SPIRVValue *SecondArgVal = transValue(II->getArgOperand(1), BB);
+
+    if (IID == Intrinsic::uadd_sat) {
+      // uadd.sat(a, b) -> res = a + b, (res >= a) ? res : MAX
+      SPIRVValue *Max =
+          transValue(Constant::getAllOnesValue(II->getType()), BB);
+      SPIRVValue *Add =
+          BM->addBinaryInst(OpIAdd, Ty, FirstArgVal, SecondArgVal, BB);
+      SPIRVValue *Cmp =
+          BM->addCmpInst(OpUGreaterThanEqual, SPVBoolTy, Add, FirstArgVal, BB);
+      return BM->addSelectInst(Cmp, Add, Max, BB);
+    }
+    if (IID == Intrinsic::usub_sat) {
+      // usub.sat(a, b) -> (a > b) ? a - b : 0
+      SPIRVValue *Sub =
+          BM->addBinaryInst(OpISub, Ty, FirstArgVal, SecondArgVal, BB);
+      SPIRVValue *Cmp = BM->addCmpInst(OpUGreaterThan, SPVBoolTy, FirstArgVal,
+                                       SecondArgVal, BB);
+      SPIRVValue *Zero = transValue(Constant::getNullValue(II->getType()), BB);
+      return BM->addSelectInst(Cmp, Sub, Zero, BB);
+    }
+
+    uint64_t NumBits = LLVMTy->getScalarSizeInBits();
+    SPIRVValue *Zero = transValue(Constant::getNullValue(II->getType()), BB);
+    SPIRVValue *Max = transValue(
+        Constant::getIntegerValue(LLVMTy, APInt::getSignedMaxValue(NumBits)),
+        BB);
+    SPIRVValue *Min = transValue(
+        Constant::getIntegerValue(LLVMTy, APInt::getSignedMinValue(NumBits)),
+        BB);
+    SPIRVValue *IsPositive =
+        BM->addCmpInst(OpSGreaterThan, SPVBoolTy, SecondArgVal, Zero, BB);
+    SPIRVValue *IsNegative =
+        BM->addCmpInst(OpSLessThan, SPVBoolTy, SecondArgVal, Zero, BB);
+
+    if (IID == Intrinsic::sadd_sat) {
+      // sadd.sat(a, b) -> if (b > 0) && a > MAX - b => overflow -> MAX
```

Comment:
Can we use this sequence:

res = a+b
if (b>0 && a>res) res = MAX
if (b<0 && a>res) res = MIN

This makes it unnecessary to calculate "MAX - b" and "MIN - b"

Reply:
Another implementation with a single comparison:

int sat_add(int a, int b) {
  int sum = a+b;

  if ((// same sign
       ~(a^b)   &
       //different sign
       (sum^a))
      // test MSB
      < 0) {
    sum = (0x7fffffff^(a>>31));
  }

  return sum;
}